# INFO370 Problem Set: Python, data frames

## Name: Gabriella Rivera


### 1 Base python
#### 1.1 Computing
Compute the following numbers, assign those to suitable names variables, and print:


1. 

In [2]:
days_per_year = 365
hours_per_day = 24
minutes_per_hour = 60
seconds_per_minute = 60
seconds_per_year = days_per_year * hours_per_day * minutes_per_hour * seconds_per_minute
print(seconds_per_year)

31536000


2.

In [3]:
age_in_years = 32.5
age_in_seconds = age_in_years * days_per_year * hours_per_day * minutes_per_hour * seconds_per_minute
print(age_in_seconds)

1024920000.0


3.

In [4]:
logical_variable = age_in_seconds > 700000000
print(logical_variable)

True


4.

In [5]:
p = 0.4
q = p * (1 - p)
print(q)

0.24


In [6]:
a = 3
x0 = 2
x = 2.5
y = (a/x0) * ((1 + (x/x0))**(-a - 1))
print(y)

0.058527663465935076


#### 1.2 Lists

1.

In [7]:
movie_list = ["Nacho Libre", "Pirates of the Caribbean", "Talk To Me", "Leave the World Behind", "Mad Max", "Just Married"]
movie_list

['Nacho Libre',
 'Pirates of the Caribbean',
 'Talk To Me',
 'Leave the World Behind',
 'Mad Max',
 'Just Married']

2.

In [8]:
three_movies = movie_list[: 3]
three_movies

['Nacho Libre', 'Pirates of the Caribbean', 'Talk To Me']

3.

In [9]:
first2 = movie_list[:2]
last2 = movie_list[4:]
first_and_last = first2 + last2
first_and_last

['Nacho Libre', 'Pirates of the Caribbean', 'Mad Max', 'Just Married']

4.

In [10]:
opposite_order = movie_list[::-1]
opposite_order

['Just Married',
 'Mad Max',
 'Leave the World Behind',
 'Talk To Me',
 'Pirates of the Caribbean',
 'Nacho Libre']

#### 1.3 Loops/dicts

1.

In [11]:
numbers = []
n = 70
for i in range(10):
    numbers.append(n + i)
numbers

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]

2.

In [12]:
sum = 0
for i in numbers:
    sum += i
mean = sum / len(numbers)
mean

74.5

3.

In [13]:
sum100 = 0
for i in range(101):
    sum100 += i
sum100

5050

4.

In [14]:
seat_dict = {'Adam' : 33, 'Ashin' : 12, 'Inukai' : 45, 'Tanaka' : 2, 'Ikki': 17}
for i in seat_dict:
    print(i + ":", seat_dict[i])
    

people = ['Adam' , 'Ashin', 'Inukai', 'Tanaka', 'Ikki']
seats = [33, 12, 45, 2, 17]
dic = {}
for i in range(4):
    dic[people[i]] = seats[i]
dic

# I'm confused as to what the exact answer is supposed to look like so I included two attemps 

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


{'Adam': 33, 'Ashin': 12, 'Inukai': 45, 'Tanaka': 2}

5.

In [15]:
l = [1, 3, 5, 7, 11, 13, 17, 19]
squares = {}
for i in l:
    squares[i] = i * i
squares

{1: 1, 3: 9, 5: 25, 7: 49, 11: 121, 13: 169, 17: 289, 19: 361}

6.

In [16]:
sum_values = 0
for i in squares.values():
    sum_values += i
sum_values

1024

#### 1.4 Functions

1.

In [17]:
import numpy as np

In [18]:
def numeric_form(time): 
    minutes_per_hour = 60
    hours = time // 100
    minutes = time % 100
    fraction_of_hours = minutes / minutes_per_hour
    result = hours + fraction_of_hours
    return result
    # return round(result, 2)  # Round to the tenths place

numeric_form(1015)
numeric_form(345)
numeric_form(-2200)

-22.0

### Basic data frames 

#### 2.1 Setup

1.

In [19]:
import pandas as pd

df = pd.read_csv('nycflights13.csv.bz2', sep="\t")

2.

In [20]:
df.info()
df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            336776 non-null  int64  
 1   month           336776 non-null  int64  
 2   day             336776 non-null  int64  
 3   dep_time        328521 non-null  float64
 4   sched_dep_time  336776 non-null  int64  
 5   dep_delay       328521 non-null  float64
 6   arr_time        328063 non-null  float64
 7   sched_arr_time  336776 non-null  int64  
 8   arr_delay       327346 non-null  float64
 9   carrier         336776 non-null  object 
 10  flight          336776 non-null  int64  
 11  tailnum         334264 non-null  object 
 12  origin          336776 non-null  object 
 13  dest            336776 non-null  object 
 14  air_time        327346 non-null  float64
 15  distance        336776 non-null  int64  
 16  hour            336776 non-null  int64  
 17  minute    

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  
2          160.0      1089     5      40  2013-01-01T10:00:00Z  
3          183.0      1576     5      45  2013-01-01T10:00:00Z  
4          116.0       762     6       0  2013-01-01T11:00:00Z  
...          ...       ...   ...     ...                   ...  
336771       NaN       213    14      55  2013-09-30T18:00:00Z  
336772       NaN       198    22       0  2013-10-01T02:00:00Z  
336773       NaN       764    12      10  2013-09-30T16:00:00Z  
336774       NaN       419    11      59  2013-09-30T15:00:00Z  
336775       NaN       431     8      40  2013-09-30T12:00:00Z  

[336776 rows x 19 columns]

#### Basic data exploration

1.

In [21]:
num_nyc_airports = df.origin.nunique()
num_nyc_airports

3

In [22]:
nyc_airports = df.origin.value_counts()
nyc_airports

origin
EWR    120835
JFK    111279
LGA    104662
Name: count, dtype: int64

2.

In [23]:
num_dest_airports_2013 = df.dest.nunique()
num_dest_airports_2013

105

3.

In [24]:
sea = df[df.dest == 'SEA']
flights_nyc_sea = sea.flight.count()
flights_nyc_sea

3923

4.

In [25]:
spokane = df[df.dest == 'GEG']
flights_nyc_spokane = spokane.flight.count()
flights_nyc_spokane

0

No, there were no flights from NYC to GEG in 2013

5.

In [26]:
dep_delays = df[df.dep_delay > 0]
sum_dep_delays = dep_delays.dep_delay.sum()
avg_dep_delay = sum_dep_delays / dep_delays.dep_delay.count()
avg_dep_delay

39.37323252771895

In [27]:
arr_delays = df[df.arr_delay > 0]
sum_arr_delays = arr_delays.arr_delay.sum()
avg_arr_delay = sum_arr_delays / df.arr_delay.count()
avg_arr_delay

16.39156733242502

The typical departure delay out of NYC during 2013 was 39.37323252771895 minutes. The typical arrival delay from NYC was 16.39156733242502 minutes. This is a suitable measure for both types of delays becuase I filtered negative delays out of the equation (flights that left and arrived early), summed the total delays for each type of delay, and computed the average by dividing by the count of delays. 

6.

Yes I did check the validity of the delay data. Like I said, I filtered out negative valued since those do not constitute delays. The sum funtion excludes NaN values by default.

#### 2.3 Let's fly to Austin!

1.

In [28]:
austin_df = df[df.dest == 'AUS']
len(austin_df.index)

2439

2.

In [29]:
austin_df

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
173     2013      1    1     905.0             905        0.0    1309.0   
439     2013      1    1    1456.0            1455        1.0    1830.0   
568     2013      1    1    1654.0            1700       -6.0    2020.0   
577     2013      1    1    1701.0            1710       -9.0    2026.0   
760     2013      1    1    2021.0            2025       -4.0    2351.0   
...      ...    ...  ...       ...             ...        ...       ...   
336360  2013      9   30    1504.0            1510       -6.0    1738.0   
336425  2013      9   30    1606.0            1605        1.0    1844.0   
336510  2013      9   30    1720.0            1720        0.0    1944.0   
336714  2013      9   30    2035.0            2029        6.0    2318.0   
336722  2013      9   30    2052.0            2054       -2.0    2306.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
173               1229       40.0      B6    1061  N281JB    JFK  AUS   
439               1813       17.0      UA    1134  N24212    EWR  AUS   
568               2034      -14.0      DL     706  N320NB    JFK  AUS   
577               2015       11.0      AA     695  N3FUAA    JFK  AUS   
760               2341       10.0      B6    1069  N203JB    JFK  AUS   
...                ...        ...     ...     ...     ...    ...  ...   
336360            1812      -34.0      UA    1483  N73256    EWR  AUS   
336425            1918      -34.0      DL     706  N389DA    JFK  AUS   
336510            2030      -46.0      AA     291  N3LBAA    JFK  AUS   
336714            2337      -19.0      UA    1039  N78524    EWR  AUS   
336722            2358      -52.0      B6    1295  N588JB    JFK  AUS   

        air_time  distance  hour  minute             time_hour  
173        263.0      1521     9       5  2013-01-01T14:00:00Z  
439        252.0      1504    14      55  2013-01-01T19:00:00Z  
568        251.0      1521    17       0  2013-01-01T22:00:00Z  
577        247.0      1521    17      10  2013-01-01T22:00:00Z  
760        255.0      1521    20      25  2013-01-02T01:00:00Z  
...          ...       ...   ...     ...                   ...  
336360     184.0      1504    15      10  2013-09-30T19:00:00Z  
336425     186.0      1521    16       5  2013-09-30T20:00:00Z  
336510     184.0      1521    17      20  2013-09-30T21:00:00Z  
336714     191.0      1504    20      29  2013-10-01T00:00:00Z  
336722     175.0      1521    20      54  2013-10-01T00:00:00Z  

[2439 rows x 19 columns]

2.

In [30]:
airlines_nyc_aus = austin_df.carrier.nunique()
airlines_nyc_aus

6

3.

In [31]:
airlines_nyc_aus = austin_df.carrier.value_counts()
airlines_nyc_aus

carrier
B6    747
UA    670
AA    365
DL    357
WN    298
9E      2
Name: count, dtype: int64

4.

In [32]:
flights2_aus = len(austin_df.index)
delay_15 = austin_df[austin_df.arr_delay > 15]
num_delayed = len(delay_15.index)
percent_delay = num_delayed / flights2_aus
percent_delay

0.24641246412464124

5.

In [33]:
numeric_form(2022)

20.366666666666667

In [37]:
# pd.set_option('display.max_rows', None)
duration = austin_df.arr_time - austin_df.dep_time
austin_df.loc[:, 'flight_duration_in_hours'] = numeric_form(duration)
austin_df

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
173     2013      1    1     905.0             905        0.0    1309.0   
439     2013      1    1    1456.0            1455        1.0    1830.0   
568     2013      1    1    1654.0            1700       -6.0    2020.0   
577     2013      1    1    1701.0            1710       -9.0    2026.0   
760     2013      1    1    2021.0            2025       -4.0    2351.0   
...      ...    ...  ...       ...             ...        ...       ...   
336360  2013      9   30    1504.0            1510       -6.0    1738.0   
336425  2013      9   30    1606.0            1605        1.0    1844.0   
336510  2013      9   30    1720.0            1720        0.0    1944.0   
336714  2013      9   30    2035.0            2029        6.0    2318.0   
336722  2013      9   30    2052.0            2054       -2.0    2306.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
173               1229       40.0      B6    1061  N281JB    JFK  AUS   
439               1813       17.0      UA    1134  N24212    EWR  AUS   
568               2034      -14.0      DL     706  N320NB    JFK  AUS   
577               2015       11.0      AA     695  N3FUAA    JFK  AUS   
760               2341       10.0      B6    1069  N203JB    JFK  AUS   
...                ...        ...     ...     ...     ...    ...  ...   
336360            1812      -34.0      UA    1483  N73256    EWR  AUS   
336425            1918      -34.0      DL     706  N389DA    JFK  AUS   
336510            2030      -46.0      AA     291  N3LBAA    JFK  AUS   
336714            2337      -19.0      UA    1039  N78524    EWR  AUS   
336722            2358      -52.0      B6    1295  N588JB    JFK  AUS   

        air_time  distance  hour  minute             time_hour  \
173        263.0      1521     9       5  2013-01-01T14:00:00Z   
439        252.0      1504    14      55  2013-01-01T19:00:00Z   
568        251.0      1521    17       0  2013-01-01T22:00:00Z   
577        247.0      1521    17      10  2013-01-01T22:00:00Z   
760        255.0      1521    20      25  2013-01-02T01:00:00Z   
...          ...       ...   ...     ...                   ...   
336360     184.0      1504    15      10  2013-09-30T19:00:00Z   
336425     186.0      1521    16       5  2013-09-30T20:00:00Z   
336510     184.0      1521    17      20  2013-09-30T21:00:00Z   
336714     191.0      1504    20      29  2013-10-01T00:00:00Z   
336722     175.0      1521    20      54  2013-10-01T00:00:00Z   

        flight_duration_in_hours  
173                     4.066667  
439                     4.233333  
568                     4.100000  
577                     3.416667  
760                     3.500000  
...                          ...  
336360                  2.566667  
336425                  2.633333  
336510                  2.400000  
336714                  3.383333  
336722                  2.900000  

[2439 rows x 20 columns]

6.

In [35]:
austin_df.flight_duration_in_hours.min()

-21.75

We might get negative values because there are cases in which a flight left an airport in NYC at 2030.0 at night, 10:30 pm, (row id# 1702 in austin_df) and got into Austin at 8.0, 8:00 am, because it had a connection with a layover, or even two connections (who knows why on earth anyone would select such a flight). Anyhow, (arr_time - dep_time) for this flight would be -2022, resulting in a negative number being used as a parameter in the numeric_form() function, resulting in a negative output.

7.

In [38]:
duration = austin_df.arr_time - austin_df.dep_time
duration = np.where(duration < 0, duration + 2400, duration)
austin_df.loc[:, 'flight_duration_in_hours'] = numeric_form(duration)
austin_df

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
173     2013      1    1     905.0             905        0.0    1309.0   
439     2013      1    1    1456.0            1455        1.0    1830.0   
568     2013      1    1    1654.0            1700       -6.0    2020.0   
577     2013      1    1    1701.0            1710       -9.0    2026.0   
760     2013      1    1    2021.0            2025       -4.0    2351.0   
...      ...    ...  ...       ...             ...        ...       ...   
336360  2013      9   30    1504.0            1510       -6.0    1738.0   
336425  2013      9   30    1606.0            1605        1.0    1844.0   
336510  2013      9   30    1720.0            1720        0.0    1944.0   
336714  2013      9   30    2035.0            2029        6.0    2318.0   
336722  2013      9   30    2052.0            2054       -2.0    2306.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
173               1229       40.0      B6    1061  N281JB    JFK  AUS   
439               1813       17.0      UA    1134  N24212    EWR  AUS   
568               2034      -14.0      DL     706  N320NB    JFK  AUS   
577               2015       11.0      AA     695  N3FUAA    JFK  AUS   
760               2341       10.0      B6    1069  N203JB    JFK  AUS   
...                ...        ...     ...     ...     ...    ...  ...   
336360            1812      -34.0      UA    1483  N73256    EWR  AUS   
336425            1918      -34.0      DL     706  N389DA    JFK  AUS   
336510            2030      -46.0      AA     291  N3LBAA    JFK  AUS   
336714            2337      -19.0      UA    1039  N78524    EWR  AUS   
336722            2358      -52.0      B6    1295  N588JB    JFK  AUS   

        air_time  distance  hour  minute             time_hour  \
173        263.0      1521     9       5  2013-01-01T14:00:00Z   
439        252.0      1504    14      55  2013-01-01T19:00:00Z   
568        251.0      1521    17       0  2013-01-01T22:00:00Z   
577        247.0      1521    17      10  2013-01-01T22:00:00Z   
760        255.0      1521    20      25  2013-01-02T01:00:00Z   
...          ...       ...   ...     ...                   ...   
336360     184.0      1504    15      10  2013-09-30T19:00:00Z   
336425     186.0      1521    16       5  2013-09-30T20:00:00Z   
336510     184.0      1521    17      20  2013-09-30T21:00:00Z   
336714     191.0      1504    20      29  2013-10-01T00:00:00Z   
336722     175.0      1521    20      54  2013-10-01T00:00:00Z   

        flight_duration_in_hours  
173                     4.066667  
439                     4.233333  
568                     4.100000  
577                     3.416667  
760                     3.500000  
...                          ...  
336360                  2.566667  
336425                  2.633333  
336510                  2.400000  
336714                  3.383333  
336722                  2.900000  

[2439 rows x 20 columns]

In [39]:
austin_df.flight_duration_in_hours.min()

2.2333333333333334

8.

Yes, my computed results are comparable to what I found after Googling flights from NYC to Austin. I get results shorter than what the flight plans show because airlines overestimate their flight times. They add some extra time to the schedule in the event of unexpected delays.

#### 2.4 Think About All This

1. I am still concerned that the flight_duration_in_hours was not computed correctly. The methodology I was able to use may have been lacking. The data seemed pretty accurate though. 

2. I can absolutely envision business-relevant analysis I could do using this data. I would love to work for an airline, not in the airport per say, but I would like to work for corporate. I love travel and I think workign with airline databases would be interesting. 

3. I don't see any ethical concerns with this analysis. I don't think anyone would check my computations if they were considering purchasing a flight.
